widget/table.go

```diff
@@ -0,0 +1,299 @@
+	NewCell        func() fyne.CanvasObject
+	UpdateCell     func(row int, col int, template fyne.CanvasObject)
+	OnCellSelected func(row int, col int)
+
+	cells                       *tableCells
+	scroll                      *ScrollContainer
+	SelectedRow, SelectedColumn int
+}
+
+// NewTable returns a new performant table widget defined by the passed functions.
+// The first returns the data size in rows and columns, second parameter is a function that returns cell
+// template objects that can be cached and the third is used to apply data at specified data location to the
+// passed template CanvasObject.
+func NewTable(size func() (int, int), create func() fyne.CanvasObject, update func(int, int, fyne.CanvasObject)) *Table {
+	t := &Table{DataSize: size, NewCell: create, UpdateCell: update, SelectedRow: -1, SelectedColumn: -1}
+	t.ExtendBaseWidget(t)
+	return t
+}
+
+// CreateRenderer returns a new renderer for the table.
+//
+// Implements: fyne.Widget
+func (t *Table) CreateRenderer() fyne.WidgetRenderer {
+	t.ExtendBaseWidget(t)
+	marker1 := canvas.NewRectangle(theme.PrimaryColor())
+	marker2 := canvas.NewRectangle(theme.PrimaryColor())
+
+	template := t.NewCell()
+	cellSize := template.MinSize().Add(fyne.NewSize(theme.Padding()*2, theme.Padding()*2))
+	t.cells = newTableCells(t, cellSize)
+	t.scroll = NewScrollContainer(t.cells)
+
+	obj := []fyne.CanvasObject{marker1, marker2, t.scroll}
+	r := &tableRenderer{t: t, rowMarker: marker1, colMarker: marker2, objects: obj, cellSize: cellSize}
+	t.scroll.onOffsetChanged = r.moveOverlay
+
+	r.Layout(t.Size())
+	return r
+}
+
+type tableRenderer struct {
+	t *Table
+
+	rowMarker, colMarker *canvas.Rectangle
+	dividers             []fyne.CanvasObject
+
+	objects  []fyne.CanvasObject
+	cellSize fyne.Size
+}
+
+func (t *tableRenderer) moveOverlay() {
+	if t.t.SelectedColumn == -1 {
+		t.colMarker.Hide()
+	} else {
+		offX := t.t.SelectedColumn*(t.cellSize.Width+1) - t.t.scroll.Offset.X
+		x1 := theme.Padding() + offX
+		x2 := x1 + t.cellSize.Width
+		if x2 < theme.Padding() || x1 > t.t.size.Width {
+			t.colMarker.Hide()
+		} else {
+			t.colMarker.Show()
+
+			left := fyne.Max(theme.Padding(), x1)
+			t.colMarker.Move(fyne.NewPos(left, 0))
+			t.colMarker.Resize(fyne.NewSize(fyne.Min(x2, t.t.size.Width)-left, theme.Padding()))
+		}
+	}
+	t.colMarker.Refresh()
+
+	if t.t.SelectedRow == -1 {
+		t.colMarker.Hide()
+	} else {
+		offY := t.t.SelectedRow*(t.cellSize.Height+1) - t.t.scroll.Offset.Y
+		y1 := theme.Padding() + offY
+		y2 := y1 + t.cellSize.Height
+		if y2 < theme.Padding() || y1 > t.t.size.Height {
+			t.rowMarker.Hide()
+		} else {
+			t.rowMarker.Show()
+
+			top := fyne.Max(theme.Padding(), y1)
+			t.rowMarker.Move(fyne.NewPos(0, top))
+			t.rowMarker.Resize(fyne.NewSize(theme.Padding(), fyne.Min(y2, t.t.size.Height)-top))
+		}
+	}
+	t.rowMarker.Refresh()
+
+	colDivs := int(math.Ceil(float64(t.t.size.Width+1) / float64(t.cellSize.Width+1)))   // +1 for div width
```

Comment:
Extracting `1` to `const tableDividerThickness = 1` would negate the need for these comments, while improving readability